query/src/main/java/com/connexta/replication/query/QueryManager.java

```diff
@@ -0,0 +1,199 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Managers for instantiating {@link QueryService} class based on configuration. */
+public class QueryManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(QueryManager.class);
+
+  /** A Facade for the various objects and services used by the query service. */
+  private final QueryServiceTools queryServiceTools;
+
+  /** Set of identifiers for the sites to be managed or empty to manage all sites. */
+  private final Set<String> sites;
+
+  /** Executor service that controls the thread for regularly checking the configured sites. */
+  private final ScheduledExecutorService executor;
+
+  /** Map of active query services keyed by site IDs. */
+  private final Map<String, QueryService> services = new HashMap<>();
+
+  /**
+   * The number of seconds to wait before reloading the site configs and updating the {@link
+   * com.connexta.replication.query.QueryService}s accordingly.
+   */
+  private long serviceRefreshPeriod;
+
+  /**
+   * Instantiate a new query manager.
+   *
+   * @param sites identifiers for the sites to be managed or empty to manage all sites
+   * @param queryServiceTools facade for various objects used by the queryService
+   */
+  public QueryManager(
+      Stream<String> sites, QueryServiceTools queryServiceTools, long serviceRefreshPeriod) {
+    this(
+        sites,
+        queryServiceTools,
+        serviceRefreshPeriod,
+        Executors.newSingleThreadScheduledExecutor());
+  }
+
+  QueryManager(
+      Stream<String> sites,
+      QueryServiceTools queryServiceTools,
+      long serviceRefreshPeriod,
+      ScheduledExecutorService executor) {
+    this.sites = sites.collect(Collectors.toSet());
+    this.queryServiceTools = queryServiceTools;
+    this.serviceRefreshPeriod = serviceRefreshPeriod;
+    this.executor = executor;
+  }
+
+  /**
+   * Initializes the query manager and kick start all query services to start polling sites that
+   * must be polled.
+   */
+  public void init() {
+    if (sites.isEmpty()) {
+      LOGGER.info("Managing queries for all configured sites.");
+    } else {
+      LOGGER.info("Managing queries for sites: {}.", sites);
+    }
+    executor.scheduleAtFixedRate(
+        this::reloadSiteConfigs, 0, serviceRefreshPeriod, TimeUnit.SECONDS);
+  }
+
+  /** Destroys the query manager and stop all currently running query services. */
+  public void destroy() {
+    LOGGER.info("Shutting down query managers");
+    executor.shutdownNow();
+    synchronized (this) {
+      for (QueryService service : services.values()) {
+        service.stop();
+      }
+      services.clear();
+    }
+  }
+
+  @VisibleForTesting
+  void reloadSiteConfigs() {
+    Stream<Site> siteStream = queryServiceTools.allSites();
+    synchronized (this) {
+      if (Thread.currentThread().isInterrupted()) {
```

Comment:
The manager will only be interrupted if it's being shutdown. The reason this block is synchronized and this check is here is because if a thread calls this method while another thread calls destroy, a new service can be created after they should all have been shutdown. This check makes sure that if we were shutdown we don't create anymore QueryServices.

Reply:
@peterhuffer he check done here doesn't affect the interruption state of the thread. Normally the executor would automatically let the thread die and create a new one. However, as Kiefer mentioned, interruptions don't magically happen and in this case, they would happen because of the shutdown done in the `destroy()` method which means the manager is shutting down.